libs/assertion/include/hpx/assertion.hpp

```diff
@@ -0,0 +1,86 @@
+//                               hpxinspect:noinclude:HPX_ASSERT
+
+//  Note: There are no include guards. This is intentional.
+
+// hpxinspect:noinclude:HPX_ASSERT
+// hpxinspect:noinclude:HPX_ASSERT_MSG
+
+#include <hpx/config.hpp>
+#include <hpx/assertion/current_function.hpp>
+#include <hpx/assertion/evaluate_assert.hpp>
+#include <hpx/assertion/source_location.hpp>
+#include <hpx/preprocessor/stringize.hpp>
+
+#include <string>
+#include <type_traits>
+
+namespace hpx { namespace assertion {
+    /// The signature for an assertion handler
+    using assertion_handler = void (*)(
+        source_location const& loc, const char* expr, std::string const& msg);
+
+    /// Set the assertion handler to be used within a program. If the handler has been
+    /// set already once, the call to this function will be ignored.
+    /// \note This function is not thread safe
+    HPX_EXPORT void set_assertion_handler(assertion_handler handler);
+}}
+
+#if defined(DOXYGEN)
+/// \def HPX_ASSERT(expr, msg)
+/// \brief This macro asserts that \a expr evaluates to true.
+///
+/// \param expr The expression to assert on. This can either be an expression
+///             that's convertible to bool or a callable which returns bool
+/// \param msg The optional message that is used to give further information if
+///             the assert fails. This should be convertible to a std::string
+///
+/// If \a expr evaluates to false, The source location and \a msg is being
+/// printed along with the expression and additional. Afterwards the program is
+/// being aborted. The assertion handler can be costumized by calling \a
+/// hpx::assertion::set_assertion_handler().
+///
+/// Asserts are enabled if \a HPX_DEBUG is set. This is the default for
+/// `CMAKE_BUILD_TYPE=Debug`
+#define HPX_ASSERT(expr)
+
+/// \see HPX_ASSERT
+#define HPX_ASSERT_MSG(expr, msg)
+#else
+/// \cond NOINTERNAL
+#define HPX_ASSERT_(expr, msg)                                                 \
+    ::hpx::assertion::detail::evaluate_assert(                                 \
+        [&]() -> bool { return !!(expr); },                                    \
+        ::hpx::assertion::source_location{                                     \
+            __FILE__, __LINE__, HPX_ASSERT_CURRENT_FUNCTION},                  \
+        HPX_PP_STRINGIZE(expr), [&]() { return msg; });                        \
+/**/
+
+// Helper macros to extract the first and the second macro paramater, the first
+// is the expression ...
+#define HPX_ASSERT_EXPR_(expr, ...) expr
+// ... and the second is the msg. msg is optional
+#define HPX_ASSERT_MSG_(expr, msg, ...) msg
+/// \encond
+
+#if defined(HPX_DEBUG)
+#define HPX_ASSERT(...)                                                        \
+    HPX_ASSERT_(HPX_ASSERT_EXPR_(__VA_ARGS__, ""),                             \
+        HPX_ASSERT_MSG_(__VA_ARGS__, std::string(), ""))                       \
+    /**/
+
+#define HPX_ASSERT_MSG(...) HPX_ASSERT(__VA_ARGS__)
```

Comment:
Should we really make `HPX_ASSERT` and `HPX_ASSERT_MSG` equivalent?

Reply:
Very good catch, but they are actually still different because of the way `HPX_ASSERT` and the helpers `HPX_ASSERT_EXPR_` and `HPX_ASSERT_MSG_` work. I.e. the message is printed correctly. It's another question if it's actually readable like this.

@sithhell do you remember why `HPX_ASSERT` or `HPX_ASSERT_MSG` would need to be variadic?

Reply:
In this implementation, they are equivalent indeed. In fact, `HPX_ASSERT_MSG` isn't necessarily needed anymore.
However, there is nothing wrong with having it implemented like this:
```
#define HPX_ASSERT(expr) HPX_ASSERT_(expr, std::string())
#define HPX_ASSERT_MSG(expr, msg) HPX_ASSERT_(expr, msg)
```
with the proposed implementation, the only thing you need to remember however is HPX_ASSERT which can take an optional message argument.

Reply:
Yes, that's what I'd like to see happening.